firewood/src/merkle.rs

```diff
@@ -1175,6 +1178,285 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
+                    return Poll::Ready(None);
+                };
+
+                let returned_key_value = match last_node.inner() {
+                    NodeType::Branch(branch) => (key, branch.value.to_owned().unwrap().to_vec()),
+                    NodeType::Leaf(leaf) => (key, leaf.1.to_vec()),
+                    NodeType::Extension(_) => todo!(),
+                };
+
+                self.key_state = IteratorState::Iterating { last_node, parents };
+
+                return Poll::Ready(Some(Ok(returned_key_value)));
+            }
+            IteratorState::Iterating {
+                last_node,
+                mut parents,
+            } => {
+                match last_node.inner() {
+                    NodeType::Branch(branch) => {
+                        // previously rendered the value from a branch node, so walk down to the first available child
+                        if let Some(child_position) =
+                            branch.children.iter().position(|&addr| addr.is_some())
+                        {
+                            let child_address = branch.children[child_position].unwrap();
+
+                            parents.push((last_node, child_position as u8)); // remember where we walked down from
+
+                            let current_node = self
+                                .merkle
+                                .get_node(child_address)
+                                .map_err(|e| api::Error::InternalError(e.into()))?;
+
+                            let found_key = key_from_parents(&parents);
+
+                            self.key_state = IteratorState::Iterating {
+                                // continue iterating from here
+                                last_node: current_node,
+                                parents,
+                            };
+
+                            found_key
+                        } else {
+                            // Branch node with no children?
+                            return Poll::Ready(Some(Err(api::Error::InternalError(Box::new(
+                                MerkleError::ParentLeafBranch,
+                            )))));
+                        }
+                    }
+                    NodeType::Leaf(leaf) => {
+                        let mut next = parents.pop();
+                        loop {
+                            match next {
+                                None => return Poll::Ready(None),
+                                Some((parent, child_position)) => {
+                                    // Assume all parents are branch nodes
+                                    let children = parent.inner().as_branch().unwrap().chd();
+
+                                    // we use wrapping_add here because the value might be u8::MAX indicating that
+                                    // we want to go down branch
+                                    let mut child_position = child_position.wrapping_add(1);
```

Comment:
If you made `child_position` an `Option<u8>`, you wouldn't have to have special values like this. Then instead, you can simply call `child_position.unwrap_or(0)` and the code would be a lot clearer. 

For the initial setup, the only change you would need is:
```rust
let mut next = parents.pop().map(|(parent, index)| (parent, index.into())));
```

Reply:
I coded this and found it far less readable. You end up with stuff like this (from key_from_parents):
`parents[0].1.unwrap() << 4 + parents[1].1.unwrap()`

I think this might be doable if a new type were added which contained an `ObjRef` and a `Option<u8>` with a bunch of helper methods, but that would be a lot more code. So I think this is a trade off between readability and volume of code.

Another possible approach I explored was to pre-add the 1 to the u8 when it's stored, leaving you with values from 0..=16 in the position field, but this unnecesarily complicates computing the current key.